sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs

```diff
@@ -16,20 +16,111 @@ namespace Azure.Iot.ModelsRepository.Fetchers
     internal class RemoteModelFetcher : IModelFetcher
     {
         private readonly HttpPipeline _pipeline;
+        private readonly ClientDiagnostics _clientDiagnostics;
         private readonly bool _tryExpanded;
 
-        public RemoteModelFetcher(ResolverClientOptions clientOptions)
+        public RemoteModelFetcher(ClientDiagnostics clientDiagnostics, ResolverClientOptions clientOptions)
         {
             _pipeline = CreatePipeline(clientOptions);
             _tryExpanded = clientOptions.DependencyResolution == DependencyResolutionOption.TryFromExpanded;
+            _clientDiagnostics = clientDiagnostics;
         }
 
         public FetchResult Fetch(string dtmi, Uri repositoryUri, CancellationToken cancellationToken = default)
         {
-            throw new NotImplementedException();
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope("RemoteModelFetcher.Fetch");
+            scope.Start();
+            try
+            {
+                Queue<string> work = PrepareWork(dtmi, repositoryUri);
+
+                string remoteFetchError = string.Empty;
+
+                while (work.Count != 0 && !cancellationToken.IsCancellationRequested)
```

Comment:
If cancellation is requested, what is the right behavior from the called method? Short circuit as we've done here? Throw an [exception](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcanceledexception?view=net-5.0)?

Reply:
Good point, I think we should throw an exception with the correct message that cancellation was requested. changing.